### PR TITLE
Kneser-Ney smoothing for better handling of unseen n-grams

### PR DESCRIPTION
**Pull request by NathanaelTamirat**
This script implements an n-gram language model with the following improvements:
- Kneser-Ney smoothing for better handling of unseen n-grams
- Expanded hyperparameter search space
- Better documentation and comments
- Improved error handling
- Slight loss reduction

### CODE REVIEW

**Comment by NathanaelTamirat** [comment]
seq_len: 3, discount: 0.10, val_loss: 2.2453, val_perplexity: 9.4429
seq_len: 3, discount: 0.30, val_loss: 2.2406, val_perplexity: 9.3986
seq_len: 3, discount: 0.50, val_loss: 2.2396, val_perplexity: 9.3896
seq_len: 3, discount: 0.70, val_loss: 2.2401, val_perplexity: 9.3945
seq_len: 3, discount: 0.90, val_loss: 2.2420, val_perplexity: 9.4125
seq_len: 4, discount: 0.10, val_loss: 2.1542, val_perplexity: 8.6207
seq_len: 4, discount: 0.30, val_loss: 2.1115, val_perplexity: 8.2603
seq_len: 4, discount: 0.50, val_loss: 2.1010, val_perplexity: 8.1740
seq_len: 4, discount: 0.70, val_loss: 2.1029, val_perplexity: 8.1895
seq_len: 4, discount: 0.90, val_loss: 2.1164, val_perplexity: 8.3013
seq_len: 5, discount: 0.10, val_loss: 2.4192, val_perplexity: 11.2372
seq_len: 5, discount: 0.30, val_loss: 2.2674, val_perplexity: 9.6539
seq_len: 5, discount: 0.50, val_loss: 2.2219, val_perplexity: 9.2252
seq_len: 5, discount: 0.70, val_loss: 2.2167, val_perplexity: 9.1767
seq_len: 5, discount: 0.90, val_loss: 2.2485, val_perplexity: 9.4733
seq_len: 6, discount: 0.10, val_loss: 2.9564, val_perplexity: 19.2287
seq_len: 6, discount: 0.30, val_loss: 2.6890, val_perplexity: 14.7163
seq_len: 6, discount: 0.50, val_loss: 2.5976, val_perplexity: 13.4317
seq_len: 6, discount: 0.70, val_loss: 2.5714, val_perplexity: 13.0842
seq_len: 6, discount: 0.90, val_loss: 2.6043, val_perplexity: 13.5222

Best model - seq_len: 4, discount: 0.50
Test loss: 2.0898, Test perplexity: 8.0835
